### PR TITLE
[Cosmetic] Character Improvements

### PR DESCRIPTION
**Pull request by SteelFall**
This "feature" is purely cosmetic.
This will give the characters in the world a bit more personality by assigning colors to them at random.

-------

What this "feature" does:

- New World will assign a random color to the starting character.

- Saving the world will also save the characters colors and loading these new saves will load the colors.

- Loading a save made prior to the addition of this feature will assign a random color to the character(s) present

-------

Screenshots:

charactercolor4_1

charactercolor4_2

charactercolor5_1

### CODE REVIEW

**Comment by BeefEX** [comment]
Screenshots please

**Comment by creationistDev** [comment]
Will upload some real soon.

**Comment by creationistDev** [comment]
Screenshots:
After creating a new world
charactercolor1_1

Saved and loaded that world after building some floor
charactercolor1_2

2 other randomly generated colors
charactercolor2

charactercolor3

**Comment by empireages** [comment]
I think this is good, but wouldn't it be better if we can "see" the helmet reflection of the player ? like modifying the alpha on this portion of the player texture.

**Comment by SteelFall** [comment]
@empireages Hmm I see what you mean. I was just not sure if it would look good as the reflection would also get the colour tint. I shall try to put in reflections and put up some screenshots of how it looks before I combine it with the rest of the PR.

**Comment by SteelFall** [comment]
As requested by @empireages, I have tweaked the player sprite to add some reflection on the player helmet.

charactercolor2_1

charactercolor2_2

Some colours may not show it as clearly as others which could be annoying but I do not know of any good way to change this.

**Comment by empireages** [comment]
It looks really good :)
I think the glass of the helmet can stay "white" and so just the cloth and the helmet border have changed colors.
what do you think ?

**Comment by SteelFall** [comment]
That would indeed look better. :laughing: 
The way that this system currently works, I might need to add another sprite renderer to the player which contains a white helmet sprite which I am sure is not optimal. Well, unless there is any other way of doing this that I am just missing ....

**Comment by empireages** [comment]
Cool that we both agreed :)
i'm very not good at graphics and things like that so i can't really help :/

**Comment by bjubes** [comment]
It's hard to color the character but keep the helmet reflection white since they are part of the same Sprite.the current Sprite system cant make them separate so the effect you want may require some tweaking of that system.

**Comment by SteelFall** [comment]
@bjubes Ah, I did not think of looking at the Sprite system too much. I was ready to resort to making a child with the helmet reflection sprite attached to it.
Just asking: Is that is still a valid solution?

**Comment by ComputerOverlord** [comment]
IDEA: Use the various characters from kenney's platformer pack.

**Comment by SteelFall** [comment]
Added a child GameObject to the character which holds a SpriteRenderer for the helmet reflection.
The reflection is also a seperate image which might be helpful for modding and creating new textures.

charactercolor3_1

charactercolor3_2

**Comment by empireages** [comment]
Nice !

**Comment by behindcurtain3** [comment]
Cool idea, looks good!

**Comment by SteelFall** [comment]
@ComputerOverlord Might be a good idea but I think that integrating other characters should be done in some PR concerning texture/sprite packs and alternate textures. If this feels completed then I could also get started on sprite packs if that is not done already.

**Comment by vogonistic** [comment]
@SteelFall Looking at the code, I think you are using tabs and are slightly off in your indentation. Can you convert each tab to 4 spaces and fix the indentation so it aligns with the code around it?

**Comment by SteelFall** [comment]
@vogonistic Will have a look at that now.

**Comment by SteelFall** [comment]
I believe that it is fixed now although I may have missed some line somewhere.

**Comment by henninglive** [comment]
Indentation is still a problem.

**Comment by SteelFall** [comment]
I am really struggling to see a difference in indentation between lines that I have added and the ones that were already there.

File: Assets/Scripts/Controllers/CharacterSpriteController.cs [hunk 1]
```diff
@@ -49,8 +49,9 @@ public void OnCharacterCreated(Character c)
         char_go.transform.SetParent(this.transform, true);
 
         SpriteRenderer sr = char_go.AddComponent<SpriteRenderer>();
-        sr.sprite = SpriteManager.current.GetSprite("Character", "p1_front");
+		sr.sprite = SpriteManager.current.GetSprite("Character", "p2_front");
```
Comment (Dormanil) [inline Assets/Scripts/Controllers/CharacterSpriteController.cs]: Old code:
screen snippet one

New code:
screen snippet two

I have marked what the indentations are made of in each snippet. Your tabs aren't turned into four spaces.

**Comment by henninglive** [comment]
Try to drag select the indentation. If the indentation character is more than one character wide than it's a tab. Most text editors have built in conversion from tabs to spaces.

**Comment by empireages** [comment]
Checkout the commited file of this PR on github itself, it's easyer to see them.

Also, in MonoDevelop, under Project>Solution Option>Source Code>Code Formating> C# source file : 
uncheck use default and check convert tab to space.

Hope this is helpful

**Comment by SteelFall** [comment]
Thank you all very much! I am currently looking at the changed files list and will go through all my lines for good measure.

**Comment by vogonistic** [comment]
What @empireages says. It's noticeable if you look at the changes tab here on github.

**Comment by empireages** [comment]
thanks @vogonistic  for clarifying  my thought :)
(I'm not very used to English)

Also can someone test my recommendation and tell if it's work well, so we put it in the wiki.

**Comment by SteelFall** [comment]
@empireages Well I am using that right now so we will see if it is working.

**Comment by vogonistic** [comment]
@SteelFall If you are using MonoDevelop, configure it the way it's shown in the images here: [CONTRIBUTING.md#Style Guidelines](https://github.com/TeamPorcupine/ProjectPorcupine/blob/master/CONTRIBUTING.md#style-guidelines)

It will automatically convert tabs to spaces, fix your indentation, brace placement and alien privacy shield.

**Comment by SteelFall** [comment]
This should be done now.

**Comment by empireages** [comment]
how I missed this markdown :/
I will too start configuring my monodevelop :)

**Comment by empireages** [comment]
yeah this is better ;)

**Comment by SteelFall** [comment]
(sighs) noticed that 2 things are still not ready :(

**Comment by vogonistic** [comment]
If you are still working on this PR, you should add `[WIP]` to the title so Quill knows not to merge it yet, then remove it when you are done.

**Comment by SteelFall** [comment]
@vogonistic Yea should have probably done that.

**Comment by SteelFall** [comment]
This PR is now [WIP] again!
What will be added:
- Characters will not be wearing helmets if the O2 level in a room passes a certain threshold.

**Comment by quill18** [comment]
@SteelFall The space suit on/off has been something I've wanted from the start (including time to equip/de-equip). So happy this is happening.

**Comment by SteelFall** [comment]
@quill18 Happy to be helpful! I am not sure if I will be able to add equipping and de-equipping without creating some messy scripts but will definitely try.

**Comment by SteelFall** [comment]
Maybe I can add a waiting time at doors so that the character checks that there is O2 on the other side of that door.

**Comment by vogonistic** [comment]
@SteelFall Is this still a `WIP` or should we consider merging it?

**Comment by SteelFall** [comment]
@vogonistic  There was one thing that I still wanted to add (the character putting on a helmet before walking through a door) but the more I am thinking about it the less I am sure if I could make a good way for the character to check the O2 Levels in the next room. I will probably add that at a later point in a new PR (aswell as adding some sound to the helmet equipping/unequipping).

**Comment by vogonistic** [comment]
I would put that in a separate PR

**Comment by SteelFall** [comment]
@vogonistic  Ok will do that!

**Comment by SteelFall** [comment]
My next PR should also make the changes here an in #444 more compatible with each other.

**Comment by vogonistic** [comment]
Merged!

**Pull request merged**